assets/src/data/model/entity-factory/extractors.js

```diff
@@ -0,0 +1,348 @@
+	return hasRawProperty( value ) ? value.raw : value;
+};
+
+/**
+ * Returns the name of a resource from the given `resourceLink`.
+ *
+ * eg. "https://api.eventespresso.com/registration" will return 'registration';
+
+ * @param {string} resourceLink
+ * @return {string} Returns the name of the resource from a provided resource
+ * link.
+ */
+export const getRelationNameFromLink = ( resourceLink ) => {
+	return camelCase( last( resourceLink.split( '/' ) ) );
+};
+
+/**
+ * Returns a plain object containing the entity field name and values from the
+ * provided entity instance
+ * @param {Object} entityInstance
+ * @return {Object} A plain object
+ */
+export const getBaseFieldsAndValuesForPersisting = ( entityInstance ) => {
+	return reduce( entityInstance.originalFieldsAndValues, (
+		fieldsAndValues,
+		fieldValue,
+		fieldName
+	) => {
+		if (
+			isEntityField( fieldName, entityInstance.schema ) &&
+			! isPrimaryKeyField( fieldName, entityInstance.schema )
+		) {
+			return fieldsAndValues[ fieldName ] = maybeConvertFromValueObject(
+				fieldValue,
+			);
+		}
+	}, {} );
+};
+
+/**
+ * Returns the primary key(s) and values for the given entityInstance
+ *
+ * @param {Object} entityInstance
+ * @return {Object} an array of values for the primary keys.
+ */
+export const getPrimaryKeyValues = ( entityInstance ) => pick(
+	entityInstance,
+	entityInstance.primaryKeys
+);
+
+/**
+ * This returns a plain object of entity fields from the schema for the entity
+ * instance (schema for fields are extracted as well).
+ *
+ * @param {Object} entityInstance
+ * @return {Object} A plain object with fields and schema properties that are
+ * entity properties.
+ */
+export const getEntityFieldsFromSchema = ( entityInstance ) => pickBy(
+	entityInstance.schema,
+	( fieldValue, fieldName ) => isEntityField(
+		fieldName,
+		entityInstance.schema
+	)
+);
+
+/**
+ * This returns a plain object of extracted primaryKey fields from the schema
+ * for the entity instance.
+ *
+ * @param {Object} entityInstance
+ * @return {Object} A plain object with fields and schema properties that
+ * 					represent primary key fields.
+ */
+export const getPrimaryKeyFieldsFromSchema = ( entityInstance ) => pickBy(
+	entityInstance.schema,
+	( fieldValue, fieldName ) => isPrimaryKeyField(
+		fieldName,
+		entityInstance.schema
+	)
+);
+
+/**
+ * Derives the default value to use for a given type.
+ *
+ * @param {string} type
+ * @return {*}  A value to use for the given type.
+ */
+export const deriveDefaultValueForType = ( type ) => {
+	if ( isArray( type ) ) {
+		return type.indexOf( 'null' ) > -1 ?
+			null :
+			deriveDefaultValueForType( type[ 0 ] );
+	}
+	switch ( type ) {
+		case 'string':
+			return '';
+		case 'number':
+		case 'integer':
+			return 0;
+		case 'null':
+		case 'object':
+			return null;
+		case 'boolean':
+		case 'bool':
+			return false;
+		case 'date-time':
+			return ( new Date() ).toISOString();
```

Comment:
mmmmmmm... i dunno... setting ALL date field defaults to NOW?
might be best to leave as `null` and let the db set its default upon persistence.

Reply:
ya this is tricky because our DateTime vo cannot be constructed with a null value.  Keep in mind that this is called by:

```js
export const getDefaultValueForField = ( fieldName, schema ) => {
	if ( schema[ fieldName ] ) {
		return schema[ fieldName ].default ?
			schema[ fieldName ].default :
			deriveDefaultValueForType( schema[ fieldName ].type );
	}
	return null;
};
```

Which means, the only time a date-time field will have a default set to now is if there is no `default` associated with that date-time in the schema.  I did a review of all our models and as far as I can tell they ALL expose a default on the schema.

Reply:
kk so what about using the `0000-01-03` default instead ? 
since that can be more reliably detected as being the equivalent of a "null" value?

Reply:
Ya that might be better I suppose. I'll implement.

Reply:
actually can't do:

```
date = new Date( 0000, 01, 03, 0, 0, 0 );
Sat Feb 03 1900 00:00:00 GMT-0500 (Eastern Standard Time)
```

after chatting in slack, we're going to leave as is.  This is consistent with all of the current `EE_Datetime_Field` fields defined server side.